### ISSUE TITLE
Make MetricRegistry.getOrAdd and MetricBuilder protected instead of private

### ISSUE THREAD

**Issue by ibuenros** [dropwizard/metrics#780]
I want to extend MetricRegistry to use slightly modified Metric implementations (e.g. will be modifying Counter, Gauge, etc.). However, the method getOrAdd and the interface MetricBuilder are both private, making this difficult.

Is there a workaround for this issue? Or can this method and interface be made public?

Thank you.

**Comment by ryantenney** [comment]
I'm afraid this won't happen in 3.1, but something like this is planned for either 3.2 or 4.0, which i hope to get out later this year.

**Comment by ibuenros** [comment]
Is there a reason we can't do the private -> protected change. It is very easy, and should not affect anything.

**Comment by wayerr** [comment]
What prevents to do like below patch for com.codahale.metrics.MetricRegistry:
```diff
--- HEAD
+++ Modified In Working Tree
@@ -50,22 +50,34 @@
 
     private final ConcurrentMap<String, Metric> metrics;
     private final List<MetricRegistryListener> listeners;
+    private final MetricBuilder<Timer> timeBuider;
+    private final MetricBuilder<Meter> meterBuider;
+    private final MetricBuilder<Histogram> histogramBuider;
+    private final MetricBuilder<Counter> counterBuider;
 
     /**
      * Creates a new {@link MetricRegistry}.
      */
     public MetricRegistry() {
-        this(new ConcurrentHashMap<String, Metric>());
+        this(new ConcurrentHashMap<String, Metric>(), new Configuration());
     }
 
+    public MetricRegistry(Configuration configuration) {
+        this(new ConcurrentHashMap<String, Metric>(), configuration);
+    }
+
     /**
      * Creates a {@link MetricRegistry} with a custom {@link ConcurrentMap} implementation for use
      * inside the registry. Call as the super-constructor to create a {@link MetricRegistry} with
      * space- or time-bounded metric lifecycles, for example.
      */
-    protected MetricRegistry(ConcurrentMap<String, Metric> metricsMap) {
+    protected MetricRegistry(ConcurrentMap<String, Metric> metricsMap, Configuration configuration) {
         this.metrics = metricsMap;
         this.listeners = new CopyOnWriteArrayList<MetricRegistryListener>();
+        this.timeBuider = configuration.timeBuider == null? MetricBuilder.TIMERS : configuration.timeBuider;
+        this.meterBuider = configuration.meterBuider == null? MetricBuilder.METERS : configuration.meterBuider;
+        this.histogramBuider = configuration.histogramBuider == null? MetricBuilder.HISTOGRAMS : configuration.histogramBuider;
+        this.counterBuider = configuration.counterBuider == null? MetricBuilder.COUNTERS : configuration.counterBuider;
     }
 
     /**
@@ -110,7 +122,7 @@
      * @return a new or pre-existing {@link Counter}
      */
     public Counter counter(String name) {
-        return getOrAdd(name, MetricBuilder.COUNTERS);
+        return getOrAdd(name, this.counterBuider);
     }
 
     /**
@@ -121,7 +133,7 @@
      * @return a new or pre-existing {@link Histogram}
      */
     public Histogram histogram(String name) {
-        return getOrAdd(name, MetricBuilder.HISTOGRAMS);
+        return getOrAdd(name, this.histogramBuider);
     }
 
     /**
@@ -132,7 +144,7 @@
      * @return a new or pre-existing {@link Meter}
      */
     public Meter meter(String name) {
-        return getOrAdd(name, MetricBuilder.METERS);
+        return getOrAdd(name, this.meterBuider);
     }
 
     /**
@@ -143,7 +155,7 @@
      * @return a new or pre-existing {@link Timer}
      */
     public Timer timer(String name) {
-        return getOrAdd(name, MetricBuilder.TIMERS);
+        return getOrAdd(name, this.timeBuider);
     }
 
     /**
@@ -312,7 +324,7 @@
             return (T) metric;
         } else if (metric == null) {
             try {
-                return register(name, builder.newMetric());
+                return register(name, builder.newMetric(new MetricBuilderConfig(name)));
             } catch (IllegalArgumentException e) {
                 final Metric added = metrics.get(name);
                 if (builder.isInstance(added)) {
@@ -397,10 +409,10 @@
     /**
      * A quick and easy way of capturing the notion of default metrics.
      */
-    private interface MetricBuilder<T extends Metric> {
+    public interface MetricBuilder<T extends Metric> {
         MetricBuilder<Counter> COUNTERS = new MetricBuilder<Counter>() {
             @Override
-            public Counter newMetric() {
+            public Counter newMetric(MetricBuilderConfig config) {
                 return new Counter();
             }
 
@@ -412,7 +424,7 @@
 
         MetricBuilder<Histogram> HISTOGRAMS = new MetricBuilder<Histogram>() {
             @Override
-            public Histogram newMetric() {
+            public Histogram newMetric(MetricBuilderConfig config) {
                 return new Histogram(new ExponentiallyDecayingReservoir());
             }
 
@@ -424,7 +436,7 @@
 
         MetricBuilder<Meter> METERS = new MetricBuilder<Meter>() {
             @Override
-            public Meter newMetric() {
+            public Meter newMetric(MetricBuilderConfig config) {
                 return new Meter();
             }
 
@@ -436,7 +448,7 @@
 
         MetricBuilder<Timer> TIMERS = new MetricBuilder<Timer>() {
             @Override
-            public Timer newMetric() {
+            public Timer newMetric(MetricBuilderConfig config) {
                 return new Timer();
             }
 
@@ -446,8 +458,64 @@
             }
         };
 
-        T newMetric();
+        T newMetric(MetricBuilderConfig config);
 
         boolean isInstance(Metric metric);
     }
+    
+    /*
+     * In future we can add some fields here, for example we can rewrite it to use with MetricName class
+     */
+    public static class MetricBuilderConfig {
+        private final String name;
+        
+        MetricBuilderConfig(String name) {
+            this.name = name;
 }
+        
+        public String getName() {
+            return name;
+        }
+    }
+    
+    public static final class Configuration {
+        private MetricBuilder<Timer> timeBuider;
+        private MetricBuilder<Meter> meterBuider;
+        private MetricBuilder<Histogram> histogramBuider;
+        private MetricBuilder<Counter> counterBuider;
+
+        public MetricBuilder<com.codahale.metrics.Timer> getTimeBuider() {
+            return timeBuider;
+        }
+
+        public void setTimeBuider(MetricBuilder<com.codahale.metrics.Timer> timeBuider) {
+            this.timeBuider = timeBuider;
+        }
+
+        public MetricBuilder<Meter> getMeterBuider() {
+            return meterBuider;
+        }
+
+        public void setMeterBuider(MetricBuilder<Meter> meterBuider) {
+            this.meterBuider = meterBuider;
+        }
+
+        public MetricBuilder<Histogram> getHistogramBuider() {
+            return histogramBuider;
+        }
+
+        public void setHistogramBuider(MetricBuilder<Histogram> histogramBuider) {
+            this.histogramBuider = histogramBuider;
+        }
+
+        public MetricBuilder<Counter> getCounterBuider() {
+            return counterBuider;
+        }
+
+        public void setCounterBuider(MetricBuilder<Counter> counterBuider) {
+            this.counterBuider = counterBuider;
+        }
+        
+        
+    }
+}

```
It allow us to use custom factories, allow use MetricName in future and also allow use external config for metrics.

**Comment by interair** [comment]
wayerr +1

**Comment by ryantenney** [comment]
Nothing at all. In fact, something very similar will ship with v4.

**Comment by wayerr** [comment]
When is planning to release v4? And where i can checkout sources for this new version? I need this feature in near future, otherwise  i will need to use ugly workarounds in production :(

**Comment by djgeary** [comment]
Another vote for this issue. 

We are currently using the the MetricsRegistry.histogram(String name) to dynamically create histograms with a name determined by data at runtime.
However we need to change from the default reservoir implementation. To do this we effectively need to copy the logic in the private getOrAdd method into our code to be able to override the Histogram creation and inject a different reservoir.

In our case making the getOrAdd and MetricBuilder **public** would simplify this and allow us just to call getOrAdd with a custom MetricBuilder.

**Comment by mzalimeni** [comment]
+1 to @djgeary's specific issue and his suggestion to make `getOrAdd` and `MetricBuilder` **public**.

Is there a downside to doing this immediately in a point release, prior to 3.2/4? This seems low risk/effort and high reward for those of us looking forward to the next version but needed an interim solution to things like alternate reservoirs.

Happy to open a pull request if you're willing.

**Comment by harelba** [comment]
+1 here as well to publicizing getOrAdd and MetricBuilder. Specific use case is to actually use a different reserviour for timers and histograms, but already encountered other use cases.

**Comment by ryantenney** [comment]
Now that 3.2 is happening, I would accept a pull request against the 3.2-development branch which made MetricBuilder public and added `meter(String, MetricBuilder<Meter>)`, `timer(String, MetricBuilder<Timer>)`, `histogram(String, MetricBuilder<Histogram>)`, `counter(String, MetricBuilder<Counter>)`.

**Comment by ryantenney** [comment]
Or, preferably, instead of making `MetricBuilder` public, introduces a new single method interface `MetricSupplier<T extends Metric>`.

**Comment by arteam** [comment]
Closed via #1023
